annorxiver_modules/corpora_comparison_helper.py

```diff
@@ -0,0 +1,411 @@
+            - (c+d)*np.log(c+d) + (a+b+c+d)*np.log(a+b+c+d)
+        )
+        log_likelihood = LL(a,b,c,d)
+
+
+        # Log Odds
+        log_ratio = float((a*d)/(b*c))
+        
+        term_data.append({
+            "lemma":term,
+            "corpus_one_a":a,
+            "corpus_two_b":b,
+            "corpus_one_c":c, 
+            "corpus_two_d":d,
+            "log_likelihood":log_likelihood,
+            "odds_ratio":log_ratio
+        })
+        
+    return pd.DataFrame.from_records(term_data)
+
+def KL_divergence(corpus_top, corpus_bottom, num_terms=100, pseudo_count=1):
+    """
+    This function calculates the KL divergence between two corpora
+    KL(D||P)
+    Arguments:
+        corpus_top - the D in the above equation
+        corpus_bottom - the P in the above equation
+        num_terms - the number of terms used to evaluate KL
+        pseudo_count - the number to add incase a term is missing from the corpus' vocab
+    """
+    term_list = list(
+        set(
+            corpus_top
+            .sort_values("count", ascending=False)
+            .head(num_terms)
+            .lemma
+            .values
+        )
+        |
+        set(
+            corpus_bottom
+            .sort_values("count", ascending=False)
+            .head(num_terms)
+            .lemma
+            .values
+        )
+    )
+    corpus_top_total = corpus_top['count'].sum()
+    corpus_bottom_total = corpus_bottom['count'].sum()
+    
+    top_term_freq = (
+        corpus_top
+        .query(f"lemma in {term_list}")
+        .assign(freq = lambda x: x['count']/corpus_top_total)
+        .sort_values("lemma")
+    )
+    
+    bottom_term_freq = (
+        corpus_bottom
+        .query(f"lemma in {term_list}")
+        .assign(freq = lambda x: x['count']/corpus_bottom_total)
+        .sort_values("lemma")
+    )
+    
+    if top_term_freq.shape[0] != bottom_term_freq.shape[0]:
+
+        missing_terms = set(term_list) -  set(bottom_term_freq.lemma.tolist())
+        if len(missing_terms) > 0:
+            bottom_term_freq = (
+                bottom_term_freq
+                .append(
+                    pd.DataFrame(missing_terms, columns=["lemma"])
+                    .assign(
+                        count=pseudo_count,
+                        freq=pseudo_count/corpus_bottom_total
+                    ),
+                )
+            )
+        
+        missing_terms =  set(term_list)- set(top_term_freq.lemma.tolist())
+        if len(missing_terms) > 0:
+            top_term_freq = (
+                top_term_freq
+                .append(
+                    pd.DataFrame(missing_terms, columns=["lemma"])
+                    .assign(
+                        count=pseudo_count,
+                        freq=pseudo_count/corpus_top_total
+                    ),
+                )
+            )
+    
+    assert top_term_freq.shape[0] == bottom_term_freq.shape[0]
+    
+    kl_div = (
```

Comment:
Looks good to me, so this is comparing the distribution of word counts between two corpora?

Reply:
>Looks good to me, so this is comparing the distribution of word counts between two corpora?

Bingo. Basically KL is comparing how similar are two corpora. based on the amount of tokens being compared. Low values mean both corpora are very similar and vice versa on high values.